### PR TITLE
Add GSoC Week 4 improvements

### PR DESCRIPTION
**Pull request by DvdMgr**
This pull request merges the features that were developed during Week 4 of GSoC. 

Main improvements:
- Progress bars for compilation and simulation running
- numpy and xarray export facilities for easy data analysis and plotting
- Move to a folder-based campaign structure for faster output file management

### CODE REVIEW

File: sem/manager.py [hunk 1]
```diff
@@ -221,3 +225,18 @@ def get_space(self, current_query, param_space, stdout_parsing_function,
 
     def __str__(self):
         return "--- Campaign info ---\n%s\n------------" % self.db
+
+    def check_repo_ok(self):
```
Comment (dizhizhou) [inline sem/manager.py]: Where is this function called? What's the usage scenario of this function?

File: examples/wifi-plotting-numpy.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+##########################
+# nWifi_values = [1, 3, 5, 7]
+# distance_values = list(range(1, 50, 10))
+# simulationTime_values = [10]
+# useRts_values = ['false', 'true']
+# mcs_values = list(range(0,8))
+# channelWidth_values = ['20', '40']
+# useShortGuardInterval_values = ['false', 'true']
+
+# Partial parameter space
```
Comment (dizhizhou) [inline examples/wifi-plotting-numpy.py]: Partial and small parameter spaces have the same setting. Is this intentional?

File: sem/manager.py (hunk 1 above)
Comment (DvdMgr) [reply sem/manager.py]: The function is to be called before performing additional simulations, in order to make sure that the same exact code is used for all simulations saved in the database. This prevents users from mixing up simulations that don't come from the same code.

File: examples/wifi-plotting-numpy.py (hunk 2 above)
Comment (DvdMgr) [reply examples/wifi-plotting-numpy.py]: No, I'll remove the partial setting!

File: sem/manager.py [hunk 3]
```diff
@@ -150,13 +152,68 @@ def run_missing_simulations(self, param_list, runs):
     # Result management #
     #####################
 
-    def get_results_as_numpy_array(self, parameter_space):
+    def get_results_as_numpy_array(self, parameter_space,
+                                   stdout_parsing_function,
```
[the remark this answers is not in the document]
Comment (DvdMgr) [reply sem/manager.py]: Sure! I'll probably get around to adding them in the next few weeks!

File: sem/manager.py [hunk 4]
```diff
@@ -150,13 +152,68 @@ def run_missing_simulations(self, param_list, runs):
...
+                                   run_averaging_function=None):
```
[the remark this answers is not in the document]
Comment (DvdMgr) [reply sem/manager.py]: For each parameter combination, we have multiple runs that yield results. This function is used to reduce the multiple runs into a single result.

I'm not really convinced about keeping this, maybe it's better to always leave all runs to the user, and let them apply the function on their own.

File: examples/wifi-plotting-numpy.py (hunk 2 above)
Comment (dizhizhou) [reply examples/wifi-plotting-numpy.py]: Thanks. Please remove all the dead codes.

File: sem/manager.py (hunk 1 above)
Comment (dizhizhou) [reply sem/manager.py]: In this case, my opinion is to only block when the ns-3 release version is different. For the case of different HEAD commit, showing a warning message and then let users to choose whether to run further simulations.

Why?
When a user uses dev version and sync with ns-3 repo everyday, it's very hard to guarantee that the latest commit will always be the same. Also, different HEAD commit doesn't mean the simulation results are different. We shouldn't make such strong assumption.

File: sem/manager.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dizhizhou) [reply sem/manager.py]: Agree. we should let users do the most of the post-processing work. The official model should focus on the core function (e.g. data collection and saving).

**Pull request merged**